packages/flutter/lib/src/material/contained_button.dart

```diff
@@ -0,0 +1,425 @@
+    if (states.contains(MaterialState.hovered))
+      return onPrimary?.withOpacity(0.08);
+    if (states.contains(MaterialState.focused) || states.contains(MaterialState.pressed))
+      return onPrimary?.withOpacity(0.24);
+    return null;
+  }
+}
+
+@immutable
+class _ContainedButtonDefaultElevation extends MaterialStateProperty<double> with Diagnosticable {
+  _ContainedButtonDefaultElevation(this.elevation);
+
+  final double elevation;
+
+  @override
+  double resolve(Set<MaterialState> states) {
+    if (states.contains(MaterialState.disabled))
+      return 0;
+    if (states.contains(MaterialState.hovered))
+      return elevation + 2;
+    if (states.contains(MaterialState.focused))
+      return elevation + 2;
+    if (states.contains(MaterialState.pressed))
+      return elevation + 6;
+    return elevation;
+  }
+}
+
+@immutable
+class _ContainedButtonDefaultMouseCursor extends MaterialStateProperty<MouseCursor> with Diagnosticable {
+  _ContainedButtonDefaultMouseCursor(this.enabledCursor, this.disabledCursor);
+
+  final MouseCursor enabledCursor;
+  final MouseCursor disabledCursor;
+
+  @override
+  MouseCursor resolve(Set<MaterialState> states) {
+    if (states.contains(MaterialState.disabled))
+      return disabledCursor;
+    return enabledCursor;
+  }
+}
+
+class _ContainedButtonWithIcon extends ContainedButton {
+  _ContainedButtonWithIcon({
+    Key key,
+    @required VoidCallback onPressed,
+    VoidCallback onLongPress,
+    ButtonStyle style,
+    FocusNode focusNode,
+    bool autofocus,
+    Clip clipBehavior,
+    @required Widget icon,
+    @required Widget label,
+  }) : assert(icon != null),
+       assert(label != null),
+       super(
+         key: key,
+         onPressed: onPressed,
+         onLongPress: onLongPress,
+         style: style,
+         focusNode: focusNode,
+         autofocus: autofocus ?? false,
+         clipBehavior: clipBehavior ?? Clip.none,
+         child: _ContainedButtonWithIconChild(icon: icon, label: label),
+      );
+
+  @override
+  ButtonStyle defaultStyleOf(BuildContext context) {
+    final EdgeInsetsGeometry scaledPadding = ButtonStyleButton.scaledPadding(
+      const EdgeInsetsDirectional.fromSTEB(12, 0, 16, 0),
+      const EdgeInsets.symmetric(horizontal: 8),
+      const EdgeInsetsDirectional.fromSTEB(8, 0, 4, 0),
+      MediaQuery.of(context, nullOk: true)?.textScaleFactor ?? 1,
+    );
+    return super.defaultStyleOf(context).copyWith(
+      padding: MaterialStateProperty.all<EdgeInsetsGeometry>(scaledPadding)
+    );
+  }
+}
+
+class _ContainedButtonWithIconChild extends StatelessWidget {
+  const _ContainedButtonWithIconChild({ Key key, this.label, this.icon }) : super(key: key);
+
+  final Widget label;
+  final Widget icon;
+
+  @override
+  Widget build(BuildContext context) {
+    final double scale = MediaQuery.of(context, nullOk: true)?.textScaleFactor ?? 1;
+    final double gap = scale <= 1 ? 8 : lerpDouble(8, 4, math.min(scale - 1, 1));
```

Comment:
I agree, in fact I'd written it that way. I changed it because the analyzer complained and I wasn't able to find a nice way around the complaint:

```
  error • The argument type 'num' can't be assigned to the parameter type 'double' •
         packages/flutter/lib/src/material/contained_button.dart:419:41 • argument_type_not_assignable
```

I believe it's complaining because clamp returns a num, not a double. Introducing another line of code didn't seem to make the statement simpler either, so ...